### PR TITLE
Update SLURM builder

### PR DESCRIPTION
**Pull request by jarlsondre**
# Summary
Updates to the `itwinai` SLURM builder. See issue #288 for more information.

Brief overview: 
- Add SLURM Builder to CLI
- Use tempdir when `--no-save-script` is active
- Add some documentation for the builder in the existing SLURM command documentation
- Add some tests for the SLURM builder
- Use `pytest`'s `tmp_path` instead of `tempfile`'s `TemporaryDirectory` wherever possible
- Remove old `slurm.sh` and `runall.sh` files from use cases
- Add training command with overridable string, removing the need for separate `slurm.py` files in EURAC and Virgo
- General clean up such as changing the type hints in the `TorchTrainer` etc.
- Possibly some more small stuff 😁

**Related issues :** #288, #316 #318, #319, #320, #324, #325

### CODE REVIEW

File: src/itwinai/cli.py [hunk 1]
```diff
@@ -163,6 +169,19 @@ def sanity_check(
         run_sanity_check(optional_deps)
 
 
+@app.command(
+    context_settings={"allow_extra_args": True, "ignore_unknown_options": True}
+)
+def generate_slurm():
```
[the remark this answers is not in the document]
Comment (jarlsondre) [reply src/itwinai/cli.py:175]: This is possible and might be better since the `Typer` help page is pretty, but since I am using the argparser, doing `-h` actually works atm and will tell you all the arguments that exist. The format will be slightly different from `Typer`'s, though. 

image

[the remark this answers is not in the document]
Comment (matbun) [reply src/itwinai/cli.py:175]: I see two limitations with this solution. First, I guess that if you use `--help` instead of `-h`, this help page will not appear -- a bit inconsistent and may generate confusion. Second, the CLI reference is not generated in the dedicated docs page

[the remark this answers is not in the document]
Comment (jarlsondre) [reply src/itwinai/cli.py:175]: Regarding the first limitation: This is possible to change by simply setting `add_help_option=False`. 

The CLI reference might not be as easy to fix. We could just add them as arguments, but there are 23 of them in total. Looking at the other `Annotated` variables they take between 1 and 8 lines each, where most seem to take 5-6. This would result in around 100-150 lines of variables. Certainly doable and might be the best fix, but certainly worth thinking about. What do you think? Should I add them? 😄

File: src/itwinai/cli.py [hunk 2]
```diff
@@ -37,7 +37,9 @@
 def generate_scalability_report(
     log_dir: Annotated[
         str,
-        typer.Option(help=("Which directory to search for the scalability metrics in.")),
+        typer.Option(
+            help=("Which directory to search for the scalability metrics in.")
+        ),
```
[the remark this answers is not in the document]
Comment (jarlsondre) [reply src/itwinai/cli.py:42]: Haha yeah I think I have my line length a bit shorter than yours. Mine doesn't use the project-wide one I think

File: src/itwinai/cli.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (matbun) [reply src/itwinai/cli.py:175]: Yes, perhaps having a test that checks that the arguments are the same wouldn't be such a bad idea, if it does not take too much time. Also, I wouldn't be to worried about the additional lines of code introduced by the arguments

[the remark this answers is not in the document]
Comment (matbun) [reply src/itwinai/cli.py:175]: Maybe there is a way to make the ArgumentParser used under the hood more strict (only for this instance), so that it expects *all* arguments to be passed and fails if unexpected arguments are passed in addition.

This way the test would be trivial

File: src/itwinai/cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matbun) [reply src/itwinai/cli.py:42]: Any change to make it consistent with the project convention?

**Comment by matbun** [comment]
I have not fully understood if the slurm config file can now be embedded into the config.yaml used to define the rest of the configuration and pipelines

**Comment by jarlsondre** [comment]
> I have not fully understood if the slurm config file can now be embedded into the config.yaml used to define the rest of the configuration and pipelines

With the current implementation it is not possible, but there are fixes that can enable this. I think it might be better to have separate files, however, since we currently have two parsers that expect different arguments. If we were to mix them, we would have to allow unrecognized arguments for both of them, which would probably make debugging significantly harder. An example is: if you write `training_command` instead of `training_cmd`, then the parsers will just ignore the argument without any warning and go automatically to the default value. This is obviously wrong but also has no warning, making it difficult to debug

**Comment by matbun** [comment]
> > I have not fully understood if the slurm config file can now be embedded into the config.yaml used to define the rest of the configuration and pipelines
> 
> With the current implementation it is not possible, but there are fixes that can enable this. I think it might be better to have separate files, however, since we currently have two parsers that expect different arguments. If we were to mix them, we would have to allow unrecognized arguments for both of them, which would probably make debugging significantly harder. An example is: if you write `training_command` instead of `training_cmd`, then the parsers will just ignore the argument without any warning and go automatically to the default value. This is obviously wrong but also has no warning, making it difficult to debug

It would be nice if we had a single config file. Apart for the obvious reason of simplicity. sometimes in the Ray configurations we may want to access some information which may be retrieved from the slurm config (e.g., num_workers in the Ray scaling config). Maybe the slurm parser could be included in the `exec-pipeline` at some point... This is when there is a real advantage for the user, IMO -- you don't even have to worry about the slurm jobscript anymore, but you just run the pipeline.

**Comment by matbun** [comment]
I have noticed that [here](https://github.com/interTwin-eu/itwinai/blob/92d3235eb72fa81a5ca21b7329a0ccb5f5debc80/src/itwinai/slurm/slurm_script_builder.py#L151) you have 

```python
rdzv_endpoint = (
    '\'$(scontrol show hostnames "$SLURM_JOB_NODELIST"'
    " | head -n 1)':29500"
)
```

whereas in the scripts we use on JSC we usually append an `i` at the end of the hostname. Example:

```bash
--rdzv_endpoint='$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)'i:29500
```

Are you sure that the communication will work? I don't know where that `i` comes from, but I saw it in all scripts for JSC

**Comment by r-sarma** [comment]
> I have noticed that [here](https://github.com/interTwin-eu/itwinai/blob/92d3235eb72fa81a5ca21b7329a0ccb5f5debc80/src/itwinai/slurm/slurm_script_builder.py#L151) you have
> 
> ```python
> rdzv_endpoint = (
>     '\'$(scontrol show hostnames "$SLURM_JOB_NODELIST"'
>     " | head -n 1)':29500"
> )
> ```
> 
> whereas in the scripts we use on JSC we usually append an `i` at the end of the hostname. Example:
> 
> ```shell
> --rdzv_endpoint='$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)'i:29500
> ```
> 
> Are you sure that the communication will work? I don't know where that `i` comes from, but I saw it in all scripts for JSC
> 
> Maybe @r-sarma knows more about that?

Yes we need this for the nodes connected with Infiniband fabric.

**Comment by matbun** [comment]
> > I have noticed that [here](https://github.com/interTwin-eu/itwinai/blob/92d3235eb72fa81a5ca21b7329a0ccb5f5debc80/src/itwinai/slurm/slurm_script_builder.py#L151) you have
> > ```python
> > rdzv_endpoint = (
> >     '\'$(scontrol show hostnames "$SLURM_JOB_NODELIST"'
> >     " | head -n 1)':29500"
> > )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > whereas in the scripts we use on JSC we usually append an `i` at the end of the hostname. Example:
> > ```shell
> > --rdzv_endpoint='$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)'i:29500
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Are you sure that the communication will work? I don't know where that `i` comes from, but I saw it in all scripts for JSC
> > Maybe @r-sarma knows more about that?
> 
> Yes we need this for the nodes connected with Infiniband fabric.

I guess that otherwise the communication will go over eth... right?

**Comment by r-sarma** [comment]
> > > I have noticed that [here](https://github.com/interTwin-eu/itwinai/blob/92d3235eb72fa81a5ca21b7329a0ccb5f5debc80/src/itwinai/slurm/slurm_script_builder.py#L151) you have
> > > ```python
> > > rdzv_endpoint = (
> > >     '\'$(scontrol show hostnames "$SLURM_JOB_NODELIST"'
> > >     " | head -n 1)':29500"
> > > )
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > whereas in the scripts we use on JSC we usually append an `i` at the end of the hostname. Example:
> > > ```shell
> > > --rdzv_endpoint='$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)'i:29500
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Are you sure that the communication will work? I don't know where that `i` comes from, but I saw it in all scripts for JSC
> > > Maybe @r-sarma knows more about that?
> > 
> > 
> > Yes we need this for the nodes connected with Infiniband fabric.
> 
> I guess that otherwise the communication will go over eth... right? @r-sarma

I am not sure. I think we got a connection error when we didn't include it. But it has been some time, so worth testing it. However, even if it works, Infiniband is preferred.

**Comment by jarlsondre** [comment]
> > > I have not fully understood if the slurm config file can now be embedded into the config.yaml used to define the rest of the configuration and pipelines
> > 
> > 
> > With the current implementation it is not possible, but there are fixes that can enable this. I think it might be better to have separate files, however, since we currently have two parsers that expect different arguments. If we were to mix them, we would have to allow unrecognized arguments for both of them, which would probably make debugging significantly harder. An example is: if you write `training_command` instead of `training_cmd`, then the parsers will just ignore the argument without any warning and go automatically to the default value. This is obviously wrong but also has no warning, making it difficult to debug
> 
> It would be nice if we had a single config file. Apart for the obvious reason of simplicity. sometimes in the Ray configurations we may want to access some information which may be retrieved from the slurm config (e.g., num_workers in the Ray scaling config). Maybe the slurm parser could be included in the `exec-pipeline` at some point... This is when there is a real advantage for the user, IMO -- you don't even have to worry about the slurm jobscript anymore, but you just run the pipeline.

I definitely see the positives of having a single file—trust me, I strongly value simplicity. Nevertheless, with the current implementation, I see more negative results than positive from this change:

Pros: 
- A single configuration file, meaning fewer files to maintain
- Avoid potentially defining the same variable in different files, such as `strategy`, which could cause contradictory declarations

Cons: 
- Increases the chance of bugs, as mistakes in the config will go unnoticed
- Decreases the ability to debug the config, as you have to manually check all variable names
- The configuration file becomes huge, thus also harder to mentally keep track of. Many topics, some of which are completely unrelated, in the same place. 
- The two parsers that would use this file are fundamentally different, having different overarching tasks, being based on different frameworks with slightly different syntax. In addition to this, the parsing even happens at different points of time in the program's lifecycle. 

In total, I would say that most—if not all—of the mental strain that is removed by having a single file is added back in with the aforementioned cons. The fact that we make the code more bug-prone in addition to this makes it seem like the trade-off is not worth it. Of course I am very open to discuss the validity of these points, but as it stands right now I am not convinced. 

Of course, it might be possible to restructure the code somehow so that the aforementioned cons are alleviated. However, I can't immediately think of what this restructuring would look like.

File: use-cases/eurac/trainer.py [hunk 3]
```diff
@@ -336,9 +338,10 @@ def create_dataloaders(self, train_dataset, validation_dataset, test_dataset):
         train_sampler = train_sampler_builder.get_sampler()
         val_sampler = val_sampler_builder.get_sampler()
 
+        batch_size = self.config.batch_size // self.strategy.global_world_size()
```
Comment (matbun) [inline use-cases/eurac/trainer.py:341]: is this really needed? It should come out of the box via the distributed sampler. Maybe the sampler is wrong?

**Review by matbun — COMMENTED** [review]
try to have your editor's formatter complying with the project's max line length

File: src/itwinai/torch/distributed.py [hunk 4]
```diff
@@ -612,6 +612,15 @@ def init(self) -> None:
...
         import deepspeed
 
+        # Removing the .put() method of the cache manager
+        # This is the same bug that was earlier removed in the generic_torch.sh script,
+        # using the sed command
+        from deepspeed.ops.transformer.inference.triton.matmul_ext import AutotuneCacheManager
+        def noop_put(self, table):
+            pass
+
+        AutotuneCacheManager.put = noop_put
```
[the remark this answers is not in the document]
Comment (jarlsondre) [reply src/itwinai/torch/distributed.py]: I agree that this is not something we want, but it's the only fix I can seem to find that is more elegant than using `sed`, which I consider to be even more bug-prone than this one. I looked at some issues on DeepSpeed's GitHub and I get the impression that this bug won't be fixed anytime soon. If you can think of something better than this, then I'd be happy to implement it.

File: use-cases/eurac/trainer.py (hunk 3 above)
Comment (jarlsondre) [reply use-cases/eurac/trainer.py:345]: I found that otherwise, the batch size became too large, causing a bunch of empty batches due to `drop_last=True`. I also found it weird, but couldn't find anything else that immediately fixed this. I will also say that EURAC has a somewhat intricate system when it comes to creating a sampler—with their own `SamplerBuilder` and so on—so this might be something that makes more sense to fix in the upcoming re-integration of this use case.

**Review by matbun — APPROVED** [review]
